server/sonar-server/src/main/java/org/sonar/server/computation/filemove/FileMoveDetectionStep.java

```diff
@@ -0,0 +1,469 @@
+            previousDbFileKey = dbFileKey;
+          }
+          res.append(score).append(separator);
+        }
+      });
+      return res.toString();
+    }
+
+    @FunctionalInterface
+    private interface ScoreMatrixVisitor {
+      void visit(String dbFileKey, String reportFileKey, int score);
+    }
+  }
+
+  @Immutable
+  private static final class Match {
+    private final String dbKey;
+    private final String reportKey;
+
+    public Match(String dbKey, String reportKey) {
+      this.dbKey = dbKey;
+      this.reportKey = reportKey;
+    }
+
+    public String getDbKey() {
+      return dbKey;
+    }
+
+    public String getReportKey() {
+      return reportKey;
+    }
+
+    @Override
+    public boolean equals(@Nullable Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      Match match = (Match) o;
+      return dbKey.equals(match.dbKey) && reportKey.equals(match.reportKey);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(dbKey, reportKey);
+    }
+
+    @Override
+    public String toString() {
+      return '{' + dbKey + "=>" + reportKey + '}';
+    }
+  }
+
+  private static class MatchesByScore implements ScoreMatrix.ScoreMatrixVisitor, Iterable<List<Match>> {
+    private final ScoreMatrix scoreMatrix;
+    private List<Match>[] matches;
+    private int totalMatches = 0;
+
+    private MatchesByScore(ScoreMatrix scoreMatrix) {
+      this.scoreMatrix = scoreMatrix;
+      this.matches = new List[max(MIN_REQUIRED_SCORE, scoreMatrix.maxScore) - MIN_REQUIRED_SCORE];
+    }
+
+    public static MatchesByScore create(ScoreMatrix scoreMatrix) {
+      MatchesByScore res = new MatchesByScore(scoreMatrix);
+      res.populate();
+      return res;
+    }
+
+    private void populate() {
+      scoreMatrix.accept(this);
+    }
+
+    @Override
+    public void visit(String dbFileKey, String reportFileKey, int score) {
+      if (!isAcceptableScore(score)) {
+        return;
+      }
+
+      int index = score - MIN_REQUIRED_SCORE - 1;
+      if (matches[index] == null) {
+        matches[index] = new ArrayList<>(1);
+      }
+      Match match = new Match(dbFileKey, reportFileKey);
+      matches[index].add(match);
+      totalMatches++;
+    }
+
+    public int getSize() {
+      return totalMatches;
+    }
+
+    @Override
+    public Iterator<List<Match>> iterator() {
+      return Arrays.asList(matches).iterator();
+    }
+
+    private static boolean isAcceptableScore(int score) {
+      return score >= MIN_REQUIRED_SCORE;
+    }
+  }
+
+  private static class ElectedMatches implements Iterable<Match> {
+    private final List<Match> matches;
+    private final Set<String> matchedFileKeys;
+    private final Predicate<Match> notAlreadyMatched = new Predicate<Match>() {
+      @Override
+      public boolean apply(@Nonnull Match input) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this "Nonnull" annotation to honor the overridden method's contract. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2638)